File: concurrent_kata/mrugger/cplusplus/kata.worker.cpp
#include <cassert>
#include "kata.worker.h"


void KataWorker::find_solutions(SequenceIterable *sequence_it)
{
  sequence_it->for_each(match_count);
}


int KataWorker::match_count(CharIterable *it)
{
  assert(it != NULL);

  int count   = 0;
  char value  = it->next();

  char c;
  while ((value > 0) && (c = it->next()))
  {
    value -= c;
    count++;
  }

  return (value == 0) ? count : 0;
}

File: concurrent_kata/mrugger/ruby/kata.worker.rb

class KataWorker

  def self.find_all_matches(iterator)
    iterator.each { |sequence_iterator| find_matches(sequence_iterator) }
  end

  def self.find_matches(iterator)
    iterator.each do |match_iterator|
      count = match_count(match_iterator)
      iterator.push_match(count) if count > 0
    end
  end

  def self.match_count(iterator)
    sum = 0
    count = 0

    value = iterator.next

    iterator.each do |char|
      sum += char
      count += 1
      break if sum >= value
    end

    sum == value ? count : 0
  end

end
